### DISCUSSION TITLE
Is there a method that can execute commands in a variable?

### DISCUSSION THREAD

**Discussion by gipsyh** [nushell/nushell#5413] [Q&A]
Is there a method that can execute commands in a variable?
just like:
```
let command = 'ls | select name'
nu -c $command
```
but `nu -c` output is string, not a list/table
is there a command can execute commands in variable and output is `nu structs`

**Comment by fdncred** [comment]
For internal commands you can do this:
`let command = (ls | select name)`
`$command`
For external commands I believe you have to do something like this:
`^$"($command)"`

**Reply by gipsyh** [reply]
I mean, `(ls | select name)` should be a string in a string variable, only can known in runtime.
```
def execute [command: string] {
    nu -c $command
}
```

**Comment by fdncred** [comment]
You can also make the variable a block, like this.
`let command = { ls | select name }`
`do $command`

**Comment by fdncred** [comment]
Maybe I'm not following what you're trying to do.
This works
```
let command = 'ls | select name'
nu -c $command
```
This works
```
def execute [command: string] {
    nu -c $command
}
```
and my suggestions work. What am I missing?

**Reply by gipsyh** [reply]
`nu -c` can work, but the output of `nu -c` is string.

is there an internal command which just like `nu -c` but the output is `nu structs` just like list, table... ?

thanks

**Comment by fdncred** [comment]
no, `nu -c` is essentially shelling out to call nushell again. i think the original intent was it could be used from bash/fish/whatever to call into nushell and get the data back.

one thing you maybe could try is ` ... | to json` and then see what happens.

**Comment by gipsyh** [comment]
I found a methord:
```
nu -c 'ls | to nuon' | from nuon
```

**Comment by DejayRezme** [comment]
That 'from nuon' is a neat trick! No reason this couldn't be an internal command.

I believe the real answer is that nushell is statically parsed. Or at least the commands and parameters are statically parsed. So if you source a script it's definition get included and following lines can be syntax checked using those definitions while parsing and before execution.

I guess pipelines are not statically parsed like parameters and always accept "any".

**Comment by wdv4758h** [comment]
Found this thread when searching for execute command in a var. Sometime I just need it to be string var and combine with other condition.

The `nu -c '$cmd | to nuon' | from nuon` seems to be a good enough workaround. (not that neat though)

**Reply by wdv4758h** [reply]
There is `run-external`, maybe we just need a `run` for nushell internal commands.

**Comment by blueforesticarus** [comment]
bumping into this trying to define a command which takes format as an input and needs to call `to $format` and `from $format`
